fakestorage/object.go

```diff
@@ -648,50 +647,135 @@ func (s *Server) downloadObject(w http.ResponseWriter, r *http.Request) {
+			`<Message>The requested range cannot be satisfied.</Message>`+
+			`<Details>%s</Details></Error>`, r.Header.Get("Range")))
+		w.Header().Set(contentTypeHeader, "application/xml; charset=UTF-8")
+	} else {
+		if obj.ContentType != "" {
+			w.Header().Set(contentTypeHeader, obj.ContentType)
+		}
+		if obj.ContentEncoding != "" {
+			w.Header().Set("Content-Encoding", obj.ContentEncoding)
+		}
 	}
+
 	w.WriteHeader(status)
 	if r.Method == http.MethodGet {
 		w.Write(content)
 	}
 }
 
-func (s *Server) handleRange(obj Object, r *http.Request) (ranged bool, start int64, lastByte int64, content []byte) {
-	if reqRange := r.Header.Get("Range"); reqRange != "" {
-		parts := strings.SplitN(reqRange, "=", 2)
-		if len(parts) == 2 && parts[0] == "bytes" {
-			rangeParts := strings.SplitN(parts[1], "-", 2)
-			if len(rangeParts) == 2 {
-				start, _ = strconv.ParseInt(rangeParts[0], 10, 64)
-				var err error
-				var end int64
-				if end, err = strconv.ParseInt(rangeParts[1], 10, 64); err != nil {
-					end = int64(len(obj.Content))
-				} else if end != math.MaxInt64 {
-					end++
-				}
-				if end > int64(len(obj.Content)) {
-					end = int64(len(obj.Content))
-				}
-				return true, start, end - 1, obj.Content[start:end]
+func (s *Server) handleRange(obj Object, r *http.Request) (ranged bool, start int64, lastByte int64, content []byte, satisfiable bool) {
+	contentLength := int64(len(obj.Content))
+	start, end, err := parseRange(r.Header.Get("Range"), contentLength)
+	if err != nil {
+		// If the range isn't valid, GCS returns all content.
+		return false, 0, 0, obj.Content, false
+	}
+	// GCS is pretty flexible when it comes to invalid ranges. A 416 http
+	// response is only returned when the range start is beyond the length of
+	// the content. Otherwise, the range is ignored.
+	switch {
+	// Invalid start. Return 416 and NO content.
+	// Examples:
+	//   Length: 40, Range: bytes=50-60
+	//   Length: 40, Range: bytes=50-
+	case start >= contentLength:
+		// This IS a ranged request, but it ISN'T satisfiable.
+		return true, 0, 0, []byte{}, false
+	// Negative range, ignore range and return all content.
+	// Examples:
+	//   Length: 40, Range: bytes=30-20
+	case end < start:
+		return false, 0, 0, obj.Content, false
+	// Return range. Clamp start and end.
+	// Examples:
+	//   Length: 40, Range: bytes=-100
+	//   Length: 40, Range: bytes=0-100
+	default:
+		if start < 0 {
+			start = 0
+		}
+		if end >= contentLength {
+			end = contentLength - 1
+		}
+		return true, start, end, obj.Content[start : end+1], true
+	}
+}
+
+// parseRange parses the range header and returns the corresponding start and
+// end indices in the content. The end index is inclusive. This function
+// doesn't validate that the start and end indices fall within the content
+// bounds. The content length is only used to handle "suffix length" and
+// range-to-end ranges.
+func parseRange(rangeHeaderValue string, contentLength int64) (start int64, end int64, err error) {
```

Comment:
probably worth it adding some tests specifically for this function, specially for the edge cases. I can take care of it later if you prefer though.

Reply:
@fsouza  If you are adding these tests, the std lib has (unexported unfortunately for our needs): https://github.com/golang/go/blob/085529bd5fa3ab508784f6a3d42f15d2dafddc65/src/net/http/fs.go#L870
which is tested here:
https://github.com/golang/go/blob/085529bd5fa3ab508784f6a3d42f15d2dafddc65/src/net/http/fs_test.go#L102

So you might be able to crib some test cases and save some time.